src/activatable/activatable.ts

```diff
@@ -0,0 +1,130 @@
+/**
+ * Created ->- Activating ->- Activated ->- Deactivating ->- Deactivated -------------->----------- Disposed
+ *                 |  |                                         |  |                                  |  |
+ *                 |  `------------------<----------------------'  `--->--- FailedToDeactivate --->---'  |
+ *                 `--------------------->--------- FailedToActivate ------------------>-----------------'
+ */
+
+enum LifeCycleState {
+  Created = 1,
+  Activating,
+  Activated,
+  Deactivating,
+  Deactivated,
+  Disposed,
+  FailedToActivate,
+  FailedToDeactivate,
+}
+
+const LifeCycleDescription = {
+  [LifeCycleState.Created]: 'has never activated before',
+  [LifeCycleState.Activating]: 'is activating',
+  [LifeCycleState.Activated]: 'is activated',
+  [LifeCycleState.Deactivating]: 'is deactivating',
+  [LifeCycleState.Deactivated]: 'is deactivated',
+  [LifeCycleState.Disposed]: 'is disposed',
+  [LifeCycleState.FailedToActivate]: 'errored while activating',
+  [LifeCycleState.FailedToDeactivate]: 'errored while deactivating',
+};
+
+export interface IActivatable {
+  isActivating: boolean;
+  isActivated: boolean;
+  isDeactivating: boolean;
+  isDeactivated: boolean;
+  isErrored: boolean;
+  activate: () => Promise<void>;
+  deactivate: () => Promise<void>;
+  disposeActivateable: () => void;
+}
+
+export abstract class Activatable implements IActivatable {
+  private _state: LifeCycleState = LifeCycleState.Created;
+
+  public get isActivating() {
```

Comment:
Just make `state` public.

Reply:
no. too easy for people to mess up

Reply:
Not if you make it non-writable, but configurable :laughing:

Reply:
look at `isDeactivated` and `isErrored` and tell me it makes sense to delete those

Reply:
They are useful.

Reply:
well none of our current code uses `isCreated` or `isDisposed` so I didn't see a point in adding those.

`IActivatable` doesn't implement `IDisposable`

Reply:
> well none of our current code uses isCreated or isDisposed so I didn't see a point in adding those.

That's because we didn't use to have such state.

Reply:
we're down to 4 is that OK? I still think it's nice to be able to do
```js
if (this.stores.foo.isActivated) {
```

rather than

```js
import { LifecycleState } from '@stoplight/lifecycle'
if (this.stores.foo.state === LifecycleState.Activated) {
```

Reply:
```
if (this.stores.foo.state === LifecycleState.Activated) {
```

Not hugely opinionated on this but we could just use string type instead of enum?

```ts
type LifecycleState = 'activated' | 'isActivating'; // | ... etc etc

export interface IActivatable {
  state: LifecycleState
}
```

```ts
if (this.stores.foo.state === 'activated') {
```

Note much of a downside afaik. The more I work with enums the more I like the idea of regular string types in certain situations.

Reply:
> Note much of a downside afaik. The more I work with enums the more I like the idea of regular string types in certain situations.

Yeah, exactly.
There is nothing wrong with `isActivated`, `isDeactivated`, but in general it's usually easier to control the flow if you have a single property you can rely on.

Reply:
(not to mention it's easy to find out what the actual state is).
If you expose `isSomething` properties only, you'd need to place a huge if else statement to learn what the current state is.

Reply:
> The more I work with enums the more I like the idea of regular string types in certain situations.

Took me a while to figure out what you meant but yeah. A Union-of-strings type is actually way better now that I think about it.